CFGpy/utils/FilesHandler.py

```
@@ -62,6 +65,51 @@ def clear_cache() -> None:
 
         return None
     
+    @staticmethod
+    def get_github_file_last_updated_date(repo_owner: str, repo_name: str, file_path: str, branch: Optional[str] = "main") -> datetime:
+        """
+        Fetch the last commit date of a file in a GitHub repository using the GitHub API.
+        """
+        url: str = f"https://api.github.com/repos/{repo_owner}/{repo_name}/commits"
+        params: dict[str, str] = {"path": file_path, "sha": branch}
+        
+        try:
+            response = requests.get(url, params=params)
+            response.raise_for_status()
+            
+            commits: dict = response.json()
+            if not commits:
+                print("No commits found for the file.")
+                return None
+            
+            last_commit_date: str = commits[0]["commit"]["committer"]["date"]
+            return datetime.datetime.strptime(last_commit_date, "%Y-%m-%dT%H:%M:%SZ")
+        
+        except requests.exceptions.RequestException as e:
+            print(f"Error fetching GitHub last updated date: {e}")
+            return None
+    
+    @staticmethod
+    def get_file_downloaded_date(*, file_path: str) -> float:
+        """
+        Retrieve the downloaded date of the file in the file system.
+        """
+        try:
+            if platform.system() == "Windows":
+                downloaded_timestamp = os.stat(file_path).st_ctime  # More reliable on Windows
+            else:
+                downloaded_timestamp = os.path.getctime(file_path)  # Works on Linux/macOS
+            return datetime.datetime.fromtimestamp(downloaded_timestamp)
+        except Exception as e:
```

catching any `Exception` is a bit risky. Do we really want to ignore all possible errors in this section?

In this case I don't think we need to catch exceptions at all, like in lines 83 and 90. In my opinion it's better to let the code fail here with a meaningful error, rather than returning `None` and letting it fail later because it doesn't behave like a `datetime` object

---

It depends how crucial this feature is, if the file exists already and is just maybe not the most updated - do we want to the code to fail or just run with a sub-ideal file and a warning. In the case that the file doesn't exists, the code will try download it anyway. 
Anyway, for now I am just adding a more information print statement to the error.

---

Maybe I'm missing something - what kinds of exception do we expect to be thrown here?

Either way, returning `None` wouldn't let the code to proceed with a sub optimal file, because it will shortly fail when trying to check if that `None` is an earlier date than github file date (line 108). I think allowing it to continue with the suboptimal file may be a good idea in some cases, for that we should catch only the specific exceptions that fit those cases, and return a default late date that will cause the comparison on line 108 to return `False` rather than to fail

---

FileNotFoundError, PermissionError, OSError and potentially a ValueError but more unlikely. I think for now I'm just going to leave the ability to continue with a suboptimal file because there shouldn't be a problem with accessing the files on git and the files aren't really going to get updated so often. If we see that there are problems later then we can add it.

---

Alright. Personally, I'd prefer letting those exceptions be thrown, see what causes them, and bypass it with `if`-`else` (if the case is such that determining the date is not possible, return some default date).

But if you choose to do it with `try`-`catch`, please catch only the specific exceptions that you know will come from a problem with accessing the date, and continuing with a default date is ok. In my opinion `catch Exception` is too broad if the except block is not guaranteed to raise anything.

---

I'm confused - I'm raising the error there, just formatted? What do you want me to change?

---

If the date cannot be determined then there will be an error, which errors do you want me to intercept and return a default date and what date do you want me to return?

---

I thought you wanted to change this to not raise an exception, and let the code continue with the old file instead. I agree this is confusing, should we continue this on Zoom?

---

If you are ok with it as is then I'm ok with it, otherwise we can continue on zoom